magento/magento2
# #22934 Improved sitemap product generation logic

### Description
Improved sitemap product generation logic for case when the 'Use Categories Path for Product URLs' is enabled.

If the 'Use Categories Path for Product URLs' is enabled the following URL generation logic for the product uses:
- if product is assigned to category (categories) use any URL with category path
- if product is not assigned to category (categories) use native product URL


### Fixed Issues
1. magento/magento2#22934: Incorrect work of "Use Categories Path for Product URLs" in sitemap generation.
2. magento/magento2#4788: Wrong sitemap product url

### Manual testing scenarios (*)
Create two categories:
- category-a 
- category-b

Create three simple products:
- simple-a, assigne it to category-a and category-b categories
- simple-b, assigne it to category-b category
- simple-c (should not be assigned to any category)

Case #1: the 'Use Categories Path for Product URLs' is enabled
1. Enable the 'Use Categories Path for Product URLs' (set to YES in Configuration -> Catalog -> Catalog -> Search Engine Optimization).
2. Generate sitemap.
3. Check the result:
```
http(s)://your-magento-instance/category-b/simple-a.html
http(s)://your-magento-instance/category-b/simple-b.html
http(s)://your-magento-instance/simple-c.html
```

Case #2: the 'Use Categories Path for Product URLs' is disabled
1. Disable the 'Use Categories Path for Product URLs' (set to NO in Configuration -> Catalog -> Catalog -> Search Engine Optimization).
2. Generate sitemap.
3. Check the result:
```
http(s)://your-magento-instance/simple-a.html
http(s)://your-magento-instance/simple-b.html
http(s)://your-magento-instance/simple-c.html
```

### Contribution checklist (*)
